### PR TITLE
Fix bad interactions between timeouts and build retires

### PR DESCRIPTION
**Pull request by dmcilvaney**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./LICENSES-AND-NOTICES/SPECS/data/licenses.json`, `./LICENSES-AND-NOTICES/SPECS/LICENSES-MAP.md`, `./LICENSES-AND-NOTICES/SPECS/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

When we queue a package to build (or test), we set a timeout (by default 8h). If the build has not finished by then we forcibly stop the build and mark it as failed.

We also support `PACKAGE_BUILD_RETRIES` and `CHECK_BUILD_RETRIES`, which will cause failed builds to re-run.

However, each time the retry was triggered the timeout would reset. For example in the buddy builds this means that a stuck package test could take 4x8=32h to build, which would exceed pipeline time limits. We want to exit gracefully with an error state so that we can generate and publish logs correctly. If the pipeline forces the timeout, it can be difficult to debug.

Instead of resetting the timeout with each retry, have all attempts share a single timeout. If the timeout is exceeded stop retrying (use `RunWithLinearBackoff()` which will take a `ctx` configured with a timeout, so we can break out early).

As part of this fix, I also noticed that the timeout handling was not cleaning up the build chroot correctly. We should not be using anything related to `panic()` for error handling, instead use `logger.Log.Fatal*()` which gives the logging library a chance to run its registered cleanup functions (ie final chroot cleanup) before exiting "gracefully".

###### Change Log  

- Package build timeout shared by all retry attempts, each invocation of `BuildAgent.BuildPacakge()` now takes a `time.Duration` instead of using the value from `BuildAgentConfig`.
- Properly clean up build chroot on timeout
  - Handle timeout logic inside the chroot.Run so we correctly exit the chroot before leaving the function, otherwise the chroot cleanup code will run from within the chroot itself and the paths will be wrong.
  - Add a new `StopAllChildProcesses()` which is like `PermanentlyStopAllChildProcesses()` but does not set the disable flag (so we can run the gpg-agent cleanup still on exit).

###### Does this affect the toolchain?  

**NO**

###### Associated issues  

- #xxxx

###### Test Methodology

(Added custom %check to words with `sleep 9h`)
- https://dev.azure.com/mariner-org/mariner/_build/results?buildId=642266&view=results

### CODE REVIEW

File: toolkit/tools/pkgworker/pkgworker.go [hunk 1]
```diff
@@ -238,6 +239,7 @@ func buildSRPMInChroot(chrootDir, rpmDirPath, toolchainDirPath, workerTar, srpmF
 	case <-time.After(timeout):
 		logger.Log.Errorf("Timeout after %v: killing all processes in chroot...", timeout)
 		shell.PermanentlyStopAllChildProcesses(unix.SIGKILL)
+		time.Sleep(5 * time.Second) // If we don't wait, the chroot mounts may still be busy
```
Comment (cwize1) [inline toolkit/tools/pkgworker/pkgworker.go:242]: This is a very unreliable way of doing this. Please try to find an alternative signal to wait on.

**Review by cwize1 — CHANGES_REQUESTED** [review]

File: toolkit/tools/pkgworker/pkgworker.go (hunk 1 above)
Comment (dmcilvaney) [reply toolkit/tools/pkgworker/pkgworker.go:242]: This is an edge-case of an edge-case type thing, but I'll see if there is a simple way to handle this better.

I'm going to say we just keep this hack if there isn't an easy fix. This failure case is very uncommon, and worst case just leaves a folder in place on the system.

I'm going to look at the gpg key service handing code, it might be failing since we call `PermanentlyStopAllChildProcesses()` and this code wasn't considered when we added the gpg stuff since its so rare.

Comment (dmcilvaney) [reply toolkit/tools/pkgworker/pkgworker.go]: Ok, figured out the underlying issue here.

File: toolkit/tools/pkgworker/pkgworker.go [hunk 2]
```diff
@@ -214,34 +215,39 @@ func buildSRPMInChroot(chrootDir, rpmDirPath, toolchainDirPath, workerTar, srpmF
...
-		shell.PermanentlyStopAllChildProcesses(unix.SIGKILL)
-		err = fmt.Errorf("build timed out after %s", timeout)
-	}
+	err = chroot.Run(func() (err error) {
+		go func() {
+			results <- buildRPMFromSRPMInChroot(srpmFileInChroot, outArch, runCheck, defines, packagesToInstall, isCCacheEnabled(ccacheManager))
+		}()
+
+		var chrootErr error = nil
+		select {
```
Comment (cwize1) [inline toolkit/tools/pkgworker/pkgworker.go:238]: Why was this moved under `chroot.Run`?

Comment (dmcilvaney) [reply toolkit/tools/pkgworker/pkgworker.go:238]: Added a quick blurb to the description, but more details:

So, we rely on `chroot.Run()` to both setup and tear down the chroot (ie restore the root dir). If we do the timeout outside the `chroot.Run()`, we will exit the function and begin teardown immediately. Unfortunately... we won't have restored the correct root yet, so all the cleanup code will be running relative to the chroot's root dir.

This means we will try to unmount stuff, but the path doesn't actually exist, so it fails.

Comment (cwize1) [reply toolkit/tools/pkgworker/pkgworker.go:238]: It would be good to add a comment in the code.

**Review by cwize1 — APPROVED** [review]

File: toolkit/tools/pkgworker/pkgworker.go (hunk 2 above)
Comment (dmcilvaney) [reply toolkit/tools/pkgworker/pkgworker.go:247]: Done

**Review by cwize1 — APPROVED** [review]
